notifier/src/main/kotlin/modules/JiraNotifier.kt

```diff
@@ -0,0 +1,140 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.notifier.modules
+
+import com.atlassian.jira.rest.client.api.JiraRestClient
+import com.atlassian.jira.rest.client.api.RestClientException
+import com.atlassian.jira.rest.client.api.domain.BasicIssue
+import com.atlassian.jira.rest.client.api.domain.Comment
+import com.atlassian.jira.rest.client.api.domain.IssueType
+import com.atlassian.jira.rest.client.api.domain.input.IssueInputBuilder
+import com.atlassian.jira.rest.client.internal.async.AsynchronousJiraRestClientFactory
+
+import java.net.URI
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.ConcurrentMap
+
+import org.ossreviewtoolkit.model.Failure
+import org.ossreviewtoolkit.model.Result
+import org.ossreviewtoolkit.model.Success
+import org.ossreviewtoolkit.model.config.JiraConfiguration
+import org.ossreviewtoolkit.utils.collectMessagesAsString
+import org.ossreviewtoolkit.utils.log
+
+class JiraNotifier(
+    val config: JiraConfiguration,
+    private val restClient: JiraRestClient = AsynchronousJiraRestClientFactory()
+        .createWithBasicHttpAuthentication(URI(config.host), config.username, config.password)
+) {
+    /** Concurrent map of all available issue types for each project */
+    private val projectIssueTypes: ConcurrentMap<String, Map<String, IssueType>> = ConcurrentHashMap()
+
+    /**
+     * A Wrapper function in order to create an issue in the given Jira instance using the Jira REST API.
+     *
+     * @param projectKey The project key of the project in Jira in which the issue should be created.
+     * @param summary The summary of the issue.
+     * @param description The description of the issue.
+     * @param issueType The issue type of the issue. The given type has to be specified in the given project in Jira.
+     * @param assignee The assignee name of the issue. The given name has to be present in the Jira instance.
+     * @param avoidDuplicates If true (default), the function will not create a new issue if there is already exactly
+     *                        one issue with the same summary, but it will add a comment to the issue with the given
+     *                        description. If false, it will always create a new issue even if there are issues with the
+     *                        same summary.
+     *
+     * @return Return a [Result] that encapsulates the created issue on success, or the caught exception on failure.
+     */
+    fun createIssue(
+        projectKey: String,
+        summary: String,
+        description: String,
+        issueType: String,
+        assignee: String? = null,
+        avoidDuplicates: Boolean = true
+    ): Result<BasicIssue> {
+        if (!projectIssueTypes.containsKey(projectKey)) {
```

Comment:
Thinking again about this map with issue types per project, I wonder whether in a typical scenario actually multiple projects are involved. I would rather assume that all issues are created in a single project only. Based on this assumption, we could get rid off the map (and probably make the API of the notifier more convenient to use) with a slightly different approach:

Rather than providing a `createIssue()` function, `JiraNotifier` could offer a function `issuesBuilder(projectKey: String)`, which is passed a project key and returns a builder object. The builder now has the `createIssue()` function and creates all issues in this project. The issue types available for this project can be loaded once when the builder is created; so there should be no longer threading issues.

This is just a suggestion, not saying it has to be implemented that way - but do you think this would make sense?

Reply:
I think there are certainly use cases for this approach, because depending on the violation/failure, the issues could possibly be created in different projects?

Reply:
Actually, my intension was to simplify the implementation. With the proposed approach, you would not have to maintain a map with projects and issue types, but load the information for a specific project once the builder for this project is created. But, as I said, just a proposal.

Reply:
BTW, in Kotlin this line should be written as `if (projectKey !in projectIssueTypes) {`.